### PR TITLE
Add support for accelerators to google_datafusion_instance

### PR DESCRIPTION
**Pull request by NickElliot**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/9070

Adds accelerator field to the data fusion api.yaml and updates the corresponding example sample test.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
datafusion: added support for `accelerators` field to `google_datafusion_instance` resource.
```

### CODE REVIEW

File: mmv1/products/datafusion/api.yaml
Comment (NickElliot) [reply mmv1/products/datafusion/api.yaml:285]: Users are supposed to be able to enable and disable their accelerators, would that still be output only?

File: mmv1/templates/terraform/examples/data_fusion_instance_full.tf.erb
Comment (NickElliot) [reply mmv1/templates/terraform/examples/data_fusion_instance_full.tf.erb]: Updated with the revised code.

File: mmv1/products/datafusion/api.yaml
Comment (shuyama1) [reply mmv1/products/datafusion/api.yaml:285]: If users are able to specify this field, then, you're right, it shouldn't be output-only. Sorry I probably got confused by the previous API issues.

File: mmv1/products/datafusion/api.yaml
Comment (shuyama1) [inline mmv1/products/datafusion/api.yaml:271]: We probably don't need this two lines here and we can see that it's not generated in the downstream https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6851-old..auto-pr-6851

File: mmv1/products/datafusion/api.yaml
Comment (NickElliot) [reply mmv1/products/datafusion/api.yaml:270]: From testing the API the "true" setup seems to be a default_from_api on the state field, with a required on the type (it doesn't error out if none is supplied, but it makes no update to the resource state regardless of supplying a state). I'll send out a revised PR momentarily.

**Review by shuyama1 — COMMENTED** [review]
LGTM overall! 
One question: Does this feature support in-place update? If so, would you mind adding a update test? Otherwise, we need to add `forcenew` (`input: true` in mmv1). Currently, update steps are not supported in the generated tests and they need to be added manually to the handwritten test. Sorry for the extra work.

**Comment by NickElliot** [comment]
Yeah, I had tested that with figuring out the default_from_api and removed the existing `input: true`, what does the update test look like?

**Comment by shuyama1** [comment]
Handwritten datafusion_instance tests locate at: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/tests/resource_data_fusion_instance_test.go.erb. You can modify an existing test to include your feature. More guidance on how to add update tests: https://googlecloudplatform.github.io/magic-modules/docs/how-to/add-handwritten-test/#update-tests. 

Let me know if you have any question implementing the test!

File: mmv1/templates/terraform/examples/data_fusion_instance_full.tf.erb
Comment (shuyama1) [inline mmv1/templates/terraform/examples/data_fusion_instance_full.tf.erb:23]: Do we want to change this to "ENABLED". As this example will also be generated in our doc, it may make more sense if the config is used to enable the feature in the example usage. But it's totally your call!

File: mmv1/products/datafusion/api.yaml
Comment (shuyama1) [inline mmv1/products/datafusion/api.yaml:287]: It doesn't look like `UNKNOWN` is a value that can be explicitly specified to by users. ([API Doc](https://cloud.google.com/data-fusion/docs/reference/rest/v1/projects.locations.instances#State_1): `UNKNOWN` - Indicates that accelerator state is currently unknown. Requests for enable, disable could be retried while in this state)
I've also tested it locally and setting `state = "UNKNOWN"` ended up causing an error (as expected). We'd likely want to remove this.

**Review by shuyama1 — COMMENTED** [review]
LGTM in general! Only have some small comments. Sorry for missing those in my previous review.

File: mmv1/templates/terraform/examples/data_fusion_instance_full.tf.erb
Comment (NickElliot) [reply mmv1/templates/terraform/examples/data_fusion_instance_full.tf.erb]: Since I needed to go ahead and do the other update I made this one too!

**Comment by ptran32** [comment]
Hi,

Any update on this PR?

**Comment by NickElliot** [comment]
We're currently waiting on an update from the service team for their API for clarification on a behavior that is breaking for Terraform functionality, once that question is answered though it shouldn't be too long until this can be merged.

**Review by shuyama1 — COMMENTED** [review]
LGTM! Thanks for adding the DSF. Should be good to go after resolving the conflicts!

**Review by shuyama1 — APPROVED** [review]
LGTM! Nice work!

File: mmv1/products/datafusion/Instance.yaml
Comment (rileykarson) [inline mmv1/products/datafusion/Instance.yaml:249]: drive-by post review comment: We could consider marking this as `required: true` as well, or mark clientside "ENABLED" default. We need more complete guidance on this, but we want to avoid `default_from_api` wherever possible- it creates worse Terraform plans, and can have surprising behaviour when performing complex updates. We still need to use it a lot of the time, though, particularly on already-released fields.

Particularly, there are often cases where making a field (in a new resource) or a subfield (of a new field) required is preferable to mapping to the minimum possible message the API allows. It creates a very clear schema, and makes the user's configuration+plan very explicit about what's happening. That's not an option for new top-level fields in existing resources or subfields in existing fields, since that'd be a breaking change.

Comment (NickElliot) [reply mmv1/products/datafusion/Instance.yaml:249]: Is there a general rule for when to approach something with "this supports the main use cases better" as opposed to just matching the API functionality? Or is it just the kind of thing that will make more sense with further experience working with the provider and adding more resources and functionality?

Comment (rileykarson) [reply mmv1/products/datafusion/Instance.yaml:249]: Right now just further experience- we need to create guidance/define a general approach, but the rules will always be pretty subjective.

My thinking here is: consider the potential for confusion here if a user has a DISABLED accelerator and the state field is not in config Terraform won't see a diff, and recreating will create a resource in a different state- vs the annoyance of typing (and realistically, copy-pasting) "ENABLED" into configs.

**Pull request merged**
